src/tasks/SingularityHandler.cpp

```diff
@@ -0,0 +1,249 @@
+
+        // singular task 
+        _task_range_s = J_svd.U.leftCols(_task_rank);
+        _projected_jacobian_s = _task_range_s.transpose() * projected_jacobian;
+        Sai2Model::OpSpaceMatrices s_matrices =
+            _robot->operationalSpaceMatrices(_projected_jacobian_s);
+        _Lambda_s = s_matrices.Lambda;
+        _Jbar_s = s_matrices.Jbar;
+        _N_s = s_matrices.N;
+    } else {
+        // includes the partial task projection handling 
+        for (int i = 1; i < _task_rank; ++i) {
+            double condition_number = J_svd.s(i) / J_svd.s(0);
+            _alpha = std::clamp((condition_number - _s_min) / (_s_max - _s_min), 0., 1.);
+
+            if (condition_number < _s_max) {
+                // non-singular task
+                _task_range_ns = J_svd.U.leftCols(i);
+                _projected_jacobian_ns = _task_range_ns.transpose() * projected_jacobian;
+                Sai2Model::OpSpaceMatrices ns_matrices =
+                    _robot->operationalSpaceMatrices(_projected_jacobian_ns);
+                _Lambda_ns = ns_matrices.Lambda;
+                _Jbar_ns = ns_matrices.Jbar;
+                _N_ns = ns_matrices.N;
+
+                // singular task 
+                _task_range_s = J_svd.U.block(0, i, J_svd.U.rows(), _task_rank - i);
+                _projected_jacobian_s = _task_range_s.transpose() * projected_jacobian;
+                Sai2Model::OpSpaceMatrices s_matrices =
+                    _robot->operationalSpaceMatrices(_projected_jacobian_s);
+                _Lambda_s = s_matrices.Lambda;
+                _Jbar_s = s_matrices.Jbar;
+                _N_s = s_matrices.N;
+                break;
+
+            } else if (i == _task_rank - 1) {
+                // fully non-singular task 
+                // non-singular task
+                _task_range_ns = J_svd.U.leftCols(_task_rank); 
+                _projected_jacobian_ns = _task_range_ns.transpose() * projected_jacobian;
+                Sai2Model::OpSpaceMatrices ns_matrices =
+                    _robot->operationalSpaceMatrices(_projected_jacobian_ns);
+                _Lambda_ns = ns_matrices.Lambda;
+                _Jbar_ns = ns_matrices.Jbar;
+                _N_ns = ns_matrices.N;
+
+                // singular task 
+                _task_range_s = MatrixXd::Zero(6, _task_rank);
+                _projected_jacobian_s = _task_range_s.transpose() * projected_jacobian;
+                _Lambda_s = MatrixXd::Zero(_task_rank, _task_rank);
+                _Jbar_s = MatrixXd::Zero(dof, _task_rank);
+                _N_s = MatrixXd::Zero(dof, dof);
+            }
+        }
+    }
+
+    // classify singularities 
+    classifySingularity(_task_range_s);
+
+    // task updates 
+    if (_sing_type == NO_SINGULARITY) {
+        _N = _N_ns;  // _N = N_ns (rank(N_ns) = task_rank)
+    } else if (_sing_type != NO_SINGULARITY) {
+        // update posture task 
+        _posture_projected_jacobian = _J_posture * _N_ns * N_prec;
```

Comment:
Why not using _projected_jacobian_s and avoid to have to compute yet another svd